### DISCUSSION TITLE
CustomLayerHost initialises but never render()s

### DISCUSSION THREAD

**Discussion by MuffinHop** [maplibre/maplibre-native#2926] [General]
I've been playing around with the gl native build, on the glfw platform, on macos.
The custom layer host is initialised but never rendered.

so `void render(const mbgl::style::CustomLayerRenderParameters&) override ` never gets called.

What my class looks like with the supporting functions.
```C
class MyCustomLayer : public mbgl::style::CustomLayerHost {
public:
    ~MyCustomLayer() {}

    void initialize() override {
        printf("ready\n");
        printf("set\n");
        printf("go\n");
    }
    float iter = 0.0f;


    void render(const mbgl::style::CustomLayerRenderParameters&) override {
        printf("iter: %f\n", iter);
        iter+=0.1f;
    }

    void contextLost() override {
        ...
    }

    void deinitialize() override {
        ...
    }
};
```

And I simply initialise and register the custom layer in the main function like so

```C
    auto myCustomLayer = std::make_unique<MyCustomLayer>();
    // Add MyCustomLayer to the map's style
    // void addLayer(std::unique_ptr<Layer>, const optional<std::string>& beforeLayerID = {});
    map.getStyle().addLayer(
        std::make_unique<mbgl::style::CustomLayer>("my-custom-layer", std::move(myCustomLayer))
        );

    backend.setMap(&map);
 ```

I assume I am missing something to render the custom layer? 

Steps I've taken so far to figure out why render isn't called:
- I tried looking at android examples for custom layers
- Replicate the unit test using Custom Layers
- looked at the mapbox api documentation for custom layers
- Searched for similar discussion online(google,github,discord,etc)

**Comment by MuffinHop** [comment]
Hmm .Looks like deinitialize() is ran before render() could be executed. Doesn't seem like documented behaviour?

**Comment by louwers** [comment]
We have switched to a different architecture, layers to not `render()` anymore, instead they emit Drawables.

@stefankarschti made progress towards a custom layer that uses Drawables. See #1754 and #1810. It is not very well documented except for some sample code and there are rough edges. It should be usable on macOS too. https://github.com/maplibre/maplibre-native/blob/main/platform/darwin/app/ExampleCustomDrawableStyleLayer.mm
